### DISCUSSION TITLE
Print variant classes not being generated/preserved with custom variant

### DISCUSSION THREAD

**Discussion by maubut** [tailwindlabs/tailwindcss#14794] [Help] [answered]
I'm trying to implement a custom print variant that works with both @media print and a class-based approach (.print-preview), but the classes are either not being generated or are being purged by Tailwind. I tried many things but nothing seem to works.

Current config;
```
// tailwind.config.js
module.exports = {
  content: ['./index.html', './src/**/*.{svelte,js,ts}'],
  safelist: ['print-mode', 'print:hidden', '*'],
  
  plugins: [
    function({ addVariant, e }) {
      addVariant('print', ({ modifySelectors, separator }) => {
        modifySelectors(({ className }) => {
          return [
            `@media print .${e(`print${separator}${className}`)}`,
            `.print-preview .${e(`print${separator}${className}`)}`,
            `.print-mode .${e(`print${separator}${className}`)}`
          ].join(',\n');
        });
      });
    }
  ]
}
```

```
<div class="print:hidden">Should be hidden in print mode</div>
<div class="print:block">Should be visible in print mode</div>

OR

<html class="print-mode"> <body><div class="print:block">Should be visible even if not in @media print since .print-mode is there</div>...
```

### Expected Behavior

- Classes with the print: prefix should be preserved during build
- These classes should apply both when printing (@media print) and when the .print-preview or .print-mode class is present on the HTML element 
-  Print-specific styles should be generated in the final CSS output

**Comment by wongjn** [comment, marked as answer]
From my debugging, it seems your variant function is not run at all. I presume it could be because its using a variant name already registered in core.

Furthermore, upon changing the variant name, it seems only one selector is generated:

```diff
-safelist: ['print-mode', 'print:hidden', '*'],
+safelist: ['print-mode', 'foo:hidden', '*'],
 …
-addVariant('print', ({ modifySelectors, separator 
+addVariant('foo', ({ modifySelectors, separator }) => {}) => {
```
```css
@media print .print\:hidden {
  display: none;
}
```

You may wish to use a more modern syntax which seems to work better: https://play.tailwindcss.com/GLlFa4Z6H0?file=config

**Reply by maubut** [reply]
Thanks buddy!
